### PR TITLE
Improve the Pulldown/hoedown warnings

### PR DESCRIPTION
**Pull request by nrc**
cc #44229

r? @QuietMisdreavus

### CODE REVIEW

**Comment by nrc** [comment]
Sample output: https://gist.github.com/nrc/e4ba8c465a5174daa616361705bf14c1

c.f., old output: https://gist.github.com/nrc/bb73f63a2a87d29ed7e10609a81ceb64

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -579,8 +583,102 @@ pub fn run(mut krate: clean::Crate,
...
+
     // And finally render the whole crate's documentation
-    cx.krate(krate)
+    let result = cx.krate(krate);
+
+    let markdown_warnings = scx.markdown_warnings.borrow();
+    if !markdown_warnings.is_empty() {
+        println!("WARNING: documentation for this crate may be rendered \
+                  differently using the new Pulldown renderer.");
+        println!("    See https://github.com/rust-lang/rust/issues/44229 for details.");
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/render.rs]: Do we want all these `println!`s to be on stdout or stderr?

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -579,8 +583,102 @@ pub fn run(mut krate: clean::Crate,
...
+            }
+        }
+    }
+
+    result
+}
+
+// A short, single-line view of `s`.
+fn concise_str(s: &str) -> String {
+    if s.contains('\n') {
```
Comment (QuietMisdreavus) [inline src/librustdoc/html/render.rs]: From the rest of the diff, i'm assuming that the text fed to this function has always been run through one of the renderers. Can either of them use Windows line endings? Does it matter?

**Review by QuietMisdreavus — COMMENTED** [review]
Couple questions

File: src/librustdoc/html/render.rs (hunk 1 above)
Comment (nrc) [reply src/librustdoc/html/render.rs]: I wanted them to be on stderr, but using `eprintln` gave no output - I'm not sure if Rustbuild or something else is swallowing stderr. I can try again to confirm I wasn't missing something, but I fear we may be stuck with stdout.

File: src/librustdoc/html/render.rs (hunk 2 above)
Comment (nrc) [reply src/librustdoc/html/render.rs]: Good catch, I can adjust that.

File: src/librustdoc/html/render.rs (hunk 1 above)
Comment (QuietMisdreavus) [reply src/librustdoc/html/render.rs]: If that's the case i won't object. I'm not sure if rustdoc has printed anything that didn't come from a panic before now.

File: src/librustdoc/html/render.rs (hunk 2 above)
Comment (QuietMisdreavus) [reply src/librustdoc/html/render.rs]: Just using something like `.lines().next()` would catch it, i assume.

**Comment by QuietMisdreavus** [comment]
From conversation on IRC, it seems like stderr is getting swallowed somewhere along the line. Possibly worth investigating, but not now. r=me pending travis.

File: src/librustdoc/html/render.rs [hunk 3]
```diff
@@ -1641,47 +1739,92 @@ fn plain_summary_line(s: Option<&str>) -> String {
...
+/// rendering between Pulldown and Hoedown.
+fn render_markdown(w: &mut fmt::Formatter,
+                   md_text: &str,
+                   span: Span,
+                   render_type: RenderType,
+                   prefix: &str,
+                   scx: &SharedContext)
+                   -> fmt::Result {
+    let hoedown_output = format!("{}", Markdown(md_text, RenderType::Hoedown));
+    // We only emit warnings if the user has opted-in to Pulldown rendering.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: I don't understand: you said you wanted warnings all the time and it was one of the reasons to delay the merge. Why this change?

File: src/librustdoc/html/render.rs [hunk 4]
```diff
@@ -1641,47 +1739,92 @@ fn plain_summary_line(s: Option<&str>) -> String {
...
+
+        pulldown_output
+    } else {
+        hoedown_output
+    };
+
     write!(w, "<div class='docblock'>{}{}</div>", prefix, output)
 }
 
+// Returns true iff s1 and s2 match, ignoring whitespace.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render.rs]: "iff"

File: src/librustdoc/html/render.rs (hunk 3 above)
Comment (nrc) [reply src/librustdoc/html/render.rs]: The core team thought that we should have this off by default until we've investigated (and probably remedied) the false positives and also communicated that these warnings are going to happen, so as not to surprise users. Sorry, for the back and forth here - my bad for not checking with the core team before asking to turn on by default.

File: src/librustdoc/html/render.rs (hunk 4 above)
Comment (nrc) [reply src/librustdoc/html/render.rs]: It's an abbreviation for 'if and only if'

File: src/librustdoc/html/render.rs [hunk 5]
```diff
@@ -579,8 +583,102 @@ pub fn run(mut krate: clean::Crate,
...
+        }
+    }
+
+    result
+}
+
+// A short, single-line view of `s`.
+fn concise_str(s: &str) -> String {
+    if s.contains('\n') {
+        return format!("{}...", s.lines().next().expect("Impossible! We just found a newline"));
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: If the first line is really long then this should probably only print the start of it like the case below.

File: src/librustdoc/html/render.rs [hunk 6]
```diff
@@ -579,8 +583,102 @@ pub fn run(mut krate: clean::Crate,
...
+    }
+    if s.len() > 70 {
+        return format!("{} ... {}", &s[..50], &s[s.len()-20..]);
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: This will panic if `50` or `s.len()-20` isn't on a character boundary.

File: src/librustdoc/html/render.rs [hunk 7]
```diff
@@ -1641,47 +1739,92 @@ fn plain_summary_line(s: Option<&str>) -> String {
...
+            .filter(|s| {
+                match *s {
+                    html_diff::Difference::NodeText { ref elem_text,
+                                                      ref opposite_elem_text,
+                                                      .. }
+                        if match_non_whitespace(elem_text, opposite_elem_text) => false,
+                    _ => true,
+                }
+            })
+            .collect::<Vec<_>>();
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: `Vec::retain` would be cleaner.

File: src/librustdoc/html/render.rs [hunk 8]
```diff
@@ -3586,3 +3729,35 @@ fn test_name_sorting() {
...
+    assert!(match_non_whitespace("a", "a"));
+    assert!(match_non_whitespace(" a ", "a"));
+    assert!(match_non_whitespace("a", "  a"));
+    assert!(match_non_whitespace("abc", "abc"));
+    assert!(match_non_whitespace("abc", " abc "));
+    assert!(match_non_whitespace("abc  ", "abc"));
+    assert!(match_non_whitespace("abc xyz", "abc xyz"));
+    assert!(match_non_whitespace("abc xyz", "abc\nxyz"));
+    assert!(match_non_whitespace("abc xyz", "abcxyz"));
+    assert!(match_non_whitespace("abcxyz", "abc xyz"));
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: Why would these be considered equal? `match_non_whitespace` is almost identical to `s1.split_whitespace().eq(s2.split_whitespace())` except for the two cases above. However, the whitepace should be handled in `html_diff` anyway as whether it matters depends on the context. For example whitespace is very important inside `<pre>` tags.

**Comment by kennytm** [comment]
@bors retry #44221.

```
[00:34:47]    Compiling rustc_save_analysis v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_save_analysis)
[00:34:54]    Compiling rustc_borrowck v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/librustc_borrowck)
No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```
